lib/Frontend/Frontend.cpp

```diff
@@ -497,58 +565,131 @@ void CompilerInstance::performSema() {
+                        DelayedParseCB);
+    if (mainIsPrimary) {
+      const auto &options = Invocation.getFrontendOptions();
+      performTypeChecking(MainFile, PersistentState.getTopLevelContext(),
+                          TypeCheckOptions, CurTUElem,
+                          options.WarnLongFunctionBodies,
+                          options.WarnLongExpressionTypeChecking,
+                          options.SolverExpressionTimeThreshold);
+    }
+    CurTUElem = MainFile.Decls.size();
+  } while (!Done);
 
-    SILParserState SILContext(TheSILModule.get());
-    unsigned CurTUElem = 0;
-    bool Done;
-    do {
-      // Pump the parser multiple times if necessary.  It will return early
-      // after parsing any top level code in a main module, or in SIL mode when
-      // there are chunks of swift decls (e.g. imports and types) interspersed
-      // with 'sil' definitions.
-      parseIntoSourceFile(MainFile, MainFile.getBufferID().getValue(), &Done,
-                          TheSILModule ? &SILContext : nullptr,
-                          &PersistentState, DelayedCB.get());
-      if (mainIsPrimary) {
-        performTypeChecking(MainFile, PersistentState.getTopLevelContext(),
-                            TypeCheckOptions, CurTUElem,
-                            options.WarnLongFunctionBodies,
-                            options.WarnLongExpressionTypeChecking,
-                            options.SolverExpressionTimeThreshold);
-      }
-      CurTUElem = MainFile.Decls.size();
-    } while (!Done);
+  Diags.setSuppressWarnings(DidSuppressWarnings);
 
-    Diags.setSuppressWarnings(DidSuppressWarnings);
-    
-    if (mainIsPrimary && !Context->hadError() &&
-        Invocation.getFrontendOptions().PCMacro) {
-      performPCMacro(MainFile, PersistentState.getTopLevelContext());
-    }
-    
-    // Playground transform knows to look out for PCMacro's changes and not
-    // to playground log them.
-    if (mainIsPrimary && !Context->hadError() &&
-        Invocation.getFrontendOptions().PlaygroundTransform)
-      performPlaygroundTransform(MainFile, Invocation.getFrontendOptions().PlaygroundHighPerformance);
-    if (!mainIsPrimary) {
-      performNameBinding(MainFile);
-    }
+  if (mainIsPrimary && !Context->hadError() &&
+      Invocation.getFrontendOptions().PCMacro) {
+    performPCMacro(MainFile, PersistentState.getTopLevelContext());
   }
 
-  // Type-check each top-level input besides the main source file.
+  // Playground transform knows to look out for PCMacro's changes and not
+  // to playground log them.
+  if (mainIsPrimary && !Context->hadError() &&
+      Invocation.getFrontendOptions().PlaygroundTransform)
+    performPlaygroundTransform(
+        MainFile, Invocation.getFrontendOptions().PlaygroundHighPerformance);
+  if (!mainIsPrimary) {
+    performNameBinding(MainFile);
+  }
+}
+
+void CompilerInstance::typeCheckTopLevelInputsExcludingMain(
```

Comment:
That's not quite what this function is doing. It's *either* type-checking every file (in WMO mode) *or* it's type-checking the primary file.

Reply:
I'm also not sure what you mean by "top-level".

Reply:
I have restructured this code, splitting into two functions.